src/slurmrestd_ops.py

```diff
@@ -0,0 +1,333 @@
+
+    def _repo(self) -> apt.DebianRepository:
+        """Return the slurm repo."""
+        ppa_url: str = "https://ppa.launchpadcontent.net/ubuntu-hpc/slurm-wlm-23.02/ubuntu"
+        sources_list: str = (
+            f"deb [signed-by={self._keyring_path}] {ppa_url} {distro.codename()} main"
+        )
+        return apt.DebianRepository.from_repo_line(sources_list)
+
+    def install(self) -> bool:
+        """Install the slurmrestd package using lib apt."""
+        # Install the key.
+        if self._keyring_path.exists():
+            self._keyring_path.unlink()
+        self._keyring_path.write_text(SLURM_PPA_KEY)
+
+        # Add the repo.
+        repositories = apt.RepositoryMapping()
+        repositories.add(self._repo())
+
+        try:
+            # Run `apt-get update`
+            apt.update()
+            apt.add_package([self._package_name, "munge", "slurm-client"])
+        except apt.PackageNotFoundError:
+            logger.error(f"{self._package_name} not found in package cache or on system")
+            return False
+        except apt.PackageError as e:
+            logger.error(f"Could not install {self._package_name}. Reason: %s", e.message)
+            return False
+        return True
+
+    def uninstall(self) -> None:
+        """Uninstall the slurmrestd package using libapt."""
+        if apt.remove_package(self._package_name):
+            logger.info(f"{self._package_name} removed from system.")
+        else:
+            logger.error(f"{self._package_name} not found on system")
+
+        repositories = apt.RepositoryMapping()
+        repositories.disable(self._repo())
+
+        # Remove the key.
+        if self._keyring_path.exists():
+            self._keyring_path.unlink()
+
+    def version(self) -> str:
+        """Return the slurmrestd version."""
+        try:
+            slurmrestd = apt.DebianPackage.from_installed_package(self._package_name)
+        except apt.PackageNotFoundError:
+            logger.error(f"{self._package_name} not found on system")
+        return slurmrestd.version.number
+
+
+class SlurmrestdOpsManager:
+    """SlurmrestdOpsManager."""
+
+    def __init__(self):
+        """Init the class."""
+        self._slurm_conf_dir = Path("/etc/slurm")
+        self._slurm_conf_path = self._slurm_conf_dir / "slurm.conf"
+
+    def install(self) -> bool:
+        """Install slurmrestd to the system."""
+        if Slurmrestd().install() is not True:
+            return False
+
+        self._create_slurmrestd_user_group()
+
+        systemd.service_enable(self._munged_systemd_service)
+        systemd.service_enable(self._slurmrestd_systemd_service)
+
+        systemd.service_stop(self._munged_systemd_service)
+
+        if not self._slurm_conf_dir.exists():
+            self._slurm_conf_dir.mkdir()
+
+        user = f"{self._slurmrestd_user_name}:{self._slurmrestd_group_name}"
+        subprocess.call(["chown", "-R", user, self._slurm_conf_dir])
```

Comment:
Any reason to not use `os.chown(...)` here? Since you're creating the slurmrestd user and group within the charm, it wouldn't be too difficult to retrieve the uid and gid of the slurmrestd user using the `pwd` and `grp` modules :smiley: